src/components/implementation/no_interface/cbboot/quarantine.c

```diff
@@ -146,140 +219,83 @@ quarantine_fork(spdid_t spdid, spdid_t source)
-	if ((d_spd = cos_spd_cntl(COS_SPD_CREATE, 0, 0, 0)) == 0) return 0;
+	/* The following, copied partly from booter.c */
+	d_spd = cos_spd_cntl(COS_SPD_CREATE, 0, 0, 0);
+	if (d_spd > MAX_NUM_SPDS) goto error;
 
 	printd("Created new spd: %d\n", d_spd);
 
-	sect = cobj_sect_get(h, 0);
+	sect = cobj_sect_get(src_hdr, 0);
 	init_daddr = sect->vaddr;
-	if (cos_spd_cntl(COS_SPD_LOCATION, d_spd, sect->vaddr, SERVICE_SIZE)) return 0;
+	if (cos_spd_cntl(COS_SPD_LOCATION, d_spd, sect->vaddr, SERVICE_SIZE)) goto error;
 	printd("Set location to %x\n", (unsigned long)sect->vaddr);
 	
-	if (cos_spd_cntl(COS_SPD_SET_FORK_ORIGIN, d_spd, source, 0)) return 0;
+	if (cos_spd_cntl(COS_SPD_SET_FORK_ORIGIN, d_spd, source, 0)) goto error;
 	printd("Set fork origin to %d\n", source);
 
-	for (j = 0 ; j < (int)h->nsect ; j++) {
-		tot += cobj_sect_size(h, j);
+	// This doesn't happen for at least this experiment, so ignore that
+	for (j = 0 ; j < (int)src_hdr->nsect ; j++) {
+		tot += cobj_sect_size(src_hdr, j);
 	}
 	if (tot > SERVICE_SIZE) {
-		if (cos_vas_cntl(COS_VAS_SPD_EXPAND, d_spd, sect->vaddr + SERVICE_SIZE, 
-				 3 * round_up_to_pgd_page(1))) {
-			printc("cbboot: could not expand VAS for component %d\n", d_spd);
-			return 0;
+		if (cos_vas_cntl(COS_VAS_SPD_EXPAND, d_spd, sect->vaddr + SERVICE_SIZE, 3 * round_up_to_pgd_page(1))) {
+			goto error;
 		}
 	}
 
 	new_sect_cbufs = cos_vect_lookup(&spd_sect_cbufs, d_spd);
 	if (!new_sect_cbufs) {
 		new_sect_cbufs = &all_spd_sect_cbufs[all_cbufs_index];
-		all_cbufs_index += h->nsect;
+		all_cbufs_index += src_hdr->nsect;
 		assert(all_cbufs_index < CBUFS_PER_PAGE * SECT_CBUF_PAGES);
-		if (cos_vect_add_id(&spd_sect_cbufs, new_sect_cbufs, d_spd) < 0) return 0;
-		if (cos_vect_add_id(&spd_sect_cbufs_header, h, d_spd) < 0) return 0;
+		if (cos_vect_add_id(&spd_sect_cbufs, new_sect_cbufs, d_spd) < 0) goto error;
+		if (cos_vect_add_id(&spd_sect_cbufs_header, src_hdr, d_spd) < 0) goto error;
 		printd("Added %d to sect_cbufs\n", d_spd);
 	}
 
 	vaddr_t prev_map = 0;
-	for (j = 0 ; j < (int)h->nsect ; j++) {
+	for (j = 0 ; j < (int)src_hdr->nsect ; j++) {
 		vaddr_t d_addr;
 		struct cbid_caddr cbm;
 		cbuf_t cbid;
 		int flags;
 		int left;
 
-		sect = cobj_sect_get(h, j);
+		sect = cobj_sect_get(src_hdr, j);
 		d_addr = sect->vaddr;
 
-		left       = cobj_sect_size(h, j);
+		left       = cobj_sect_size(src_hdr, j);
 		/* previous section overlaps with this one, don't remap! */
 		if (round_to_page(d_addr) == prev_map) {
 			left -= (prev_map + PAGE_SIZE - d_addr);
 			d_addr = prev_map + PAGE_SIZE;
 		}
-		printd("Mapping section %d @ %x with %d bytes\n", j, (unsigned long)d_addr, left);
 		if (left > 0) {
 			left = round_up_to_page(left);
 			prev_map = d_addr;
 
 			if (sect->flags & COBJ_SECT_WRITE) flags = MAPPING_RW;
 			else flags = MAPPING_READ;
-			flags |= 2; /* no valloc */
-
+			flags |= MAPPING_NO_VALLOC;
 			cbm = old_sect_cbufs[j];
-			/* if RW, don't want to share, need new mem */
```

Comment:
Hm... I don't know what you mean by this being a problem, can you clarify? Right now, we create a new cbuf for all the memory. This will change in the future but isn't part of this PR.